### PR TITLE
[TSan] Add positive test for TSan + Dispatch on Linux

### PR DESCRIPTION
**Pull request by yln**
1) Enable tests that use `import Dispatch` on Linux. Add substitution
       `%import-libdispatch` that needs to be used for all cross-platform
       tests (i.e., tests that are intended to be run on other platforms
       than Darwin) that do `import Dispatch` or enable thread sanitizer.
    
2) Make sure as many existing Dispatch and TSan tests as possible run on
       Linux. Mark tests that would require substantial work with
       `UNSUPPORTED: OS=linux-gnu`.
    
3) Add integration-style Swift test that shows that TSan finds a simple
       race when using `Dispatch.async` incorrectly. A more complete test
       suite for TSan's libdispatch support lives on the LLVM/compiler-rt
       side.
    
rdar://problem/49177535

### CODE REVIEW

File: test/stdlib/Dispatch.swift
Comment (yln) [inline test/stdlib/Dispatch.swift]:
```
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:14:42: error: type 'DispatchSource' has no member 'ProcessEvent'
  expectEqual(2147483648, DispatchSource.ProcessEvent.exit.rawValue)
                          ~~~~~~~~~~~~~~ ^~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:87:13: error: use of unresolved identifier 'mach_timebase_info_data_t'
        var info = mach_timebase_info_data_t(numer: 1, denom: 1)
                   ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:88:2: error: use of unresolved identifier 'mach_timebase_info'
        mach_timebase_info(&info)
        ^~~~~~~~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:95:46: error: use of unresolved identifier 'NSEC_PER_SEC'
        time = DispatchTime(uptimeNanoseconds: 15 * NSEC_PER_SEC)
                                                    ^~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:96:43: error: use of unresolved identifier 'NSEC_PER_SEC'
        expectEqual(time.uptimeNanoseconds, 15 * NSEC_PER_SEC)
                                                 ^~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:148:14: error: use of unresolved identifier 'NSEC_PER_SEC'
        expectEqual(NSEC_PER_SEC, diffNanos)
                    ^~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:153:66: error: use of unresolved identifier 'O_RDONLY'
        let chan = DispatchIO(type: .random, path: "_REL_PATH_", oflag: O_RDONLY, mode: 0, queue: q, cleanupHandler: { (error) in })
                                                                        ^~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:200:36: error: argument type 'DispatchTimeInterval' does not conform to expected type 'Equatable'
                expectEqual(DispatchTimeInterval.seconds(i), DispatchTimeInterval.milliseconds(i * 1000))
                                                 ^
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:201:36: error: argument type 'DispatchTimeInterval' does not conform to expected type 'Equatable'
                expectEqual(DispatchTimeInterval.milliseconds(i), DispatchTimeInterval.microseconds(i * 1000))
                                                 ^
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:202:36: error: argument type 'DispatchTimeInterval' does not conform to expected type 'Equatable'
                expectEqual(DispatchTimeInterval.microseconds(i), DispatchTimeInterval.nanoseconds(i * 1000))
                                                 ^
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:228:46: error: binary operator '!=' cannot be applied to two 'DispatchTimeInterval' operands
        expectTrue(DispatchTimeInterval.seconds(10) != DispatchTimeInterval.never);
                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:228:46: note: binary operator '!=' cannot be synthesized for enums with associated values
        expectTrue(DispatchTimeInterval.seconds(10) != DispatchTimeInterval.never);
                                                    ^
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:229:40: error: binary operator '!=' cannot be applied to two 'DispatchTimeInterval' operands
        expectTrue(DispatchTimeInterval.never != DispatchTimeInterval.seconds(10));
                   ~~~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:229:40: note: binary operator '!=' cannot be synthesized for enums with associated values
        expectTrue(DispatchTimeInterval.never != DispatchTimeInterval.seconds(10));
                                              ^
```

File: test/stdlib/DispatchTypes.swift
Comment (yln) [inline test/stdlib/DispatchTypes.swift]:
```
/home/yln/work/swift-source/swift/test/stdlib/DispatchTypes.swift:27:45: error: use of unresolved identifier 'mach_port_t'
_ = DispatchSource.makeMachSendSource(port: mach_port_t(0), eventMask: [])
                                            ^~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/DispatchTypes.swift:28:48: error: use of unresolved identifier 'mach_port_t'
_ = DispatchSource.makeMachReceiveSource(port: mach_port_t(0))
                                               ^~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/DispatchTypes.swift:29:20: error: type 'DispatchSource' has no member 'makeMemoryPressureSource'
_ = DispatchSource.makeMemoryPressureSource(eventMask: [])
    ~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/DispatchTypes.swift:30:20: error: type 'DispatchSource' has no member 'makeProcessSource'; did you mean 'makeWriteSource'?
_ = DispatchSource.makeProcessSource(identifier: 0, eventMask: [])
    ~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~
                   makeWriteSource
Dispatch.DispatchSource:32:23: note: 'makeWriteSource' declared here
    public class func makeWriteSource(fileDescriptor: Int32, queue: DispatchQueue? = nil) -> DispatchSourceWrite
                      ^
/home/yln/work/swift-source/swift/test/stdlib/DispatchTypes.swift:32:45: error: use of unresolved identifier 'SIGINT'
_ = DispatchSource.makeSignalSource(signal: SIGINT)
                                            ^~~~~~
/home/yln/work/swift-source/swift/test/stdlib/DispatchTypes.swift:34:20: error: type 'DispatchSource' has no member 'makeFileSystemObjectSource'
_ = DispatchSource.makeFileSystemObjectSource(fileDescriptor: 0, eventMask: [])
    ~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: test/ClangImporter/Dispatch_test.swift
Comment (compnerd) [reply test/ClangImporter/Dispatch_test.swift]: Sounds like `-DDEPLOYMENT_TARGET_LINUX` may be missing?

File: test/stdlib/Dispatch.swift
Comment (compnerd) [reply test/stdlib/Dispatch.swift]: Sounds like `-DDEPLOYMENT_TARGET_LINUX` may be missing?

File: test/stdlib/DispatchTypes.swift
Comment (compnerd) [reply test/stdlib/DispatchTypes.swift]: Sounds like `-DDEPLOYMENT_TARGET_LINUX` may be missing?

File: test/stdlib/Dispatch.swift
Comment (yln) [reply test/stdlib/Dispatch.swift]: I searched for `DEPLOYMENT_TARGET_LINUX` and it seems to be a swift-corelibs-foundation option. Currently I am building simply via `./swift/utils/build-script  --release-debuginfo  --libdispatch --install-libdispatch`, e.g. only worrying about libdispatch at the moment.

Do you have an idea how to make `import Dispatch` work in the tests without copying things into the <swift-build> directory?

File: test/Sanitizers/tsan-libdispatch.swift
Comment (jrose-apple) [reply test/Sanitizers/tsan-libdispatch.swift]: Feel free to add a substitution for it.

File: test/stdlib/Dispatch.swift
Comment (jrose-apple) [reply test/stdlib/Dispatch.swift]: Add search paths using lit, I think.

Comment (compnerd) [reply test/stdlib/Dispatch.swift]: For the link, you will need to add `-L` for the search paths.  For the runtime, you will need to add `LD_LIBRARY_PATH` on Linux, `DYLD_LIBRARY_PATH` on macOS, and `PATH` on Windows.

**Comment by yln** [comment]
@swift-ci Please test

File: test/lit.cfg
Comment (compnerd) [inline test/lit.cfg]: I think that you should make this more resilient and ensure that dispatch, swiftDispatch, and the swift module are available before declaring the feature as being available.

File: test/lit.cfg
Comment (compnerd) [inline test/lit.cfg]: This is really unfortunate.  I know that I do not build the same way that the build-script sets things up (even ignoring that the paths are too long that way for Windows).  @davezarzycki may also be building differently.  Can you add an option to CMake that is wired down to the configuration instead?

File: test/lit.cfg
Comment (compnerd) [inline test/lit.cfg]: I think that it is nicer if you wrote this as `run_ptrsize == 64`

File: test/stdlib/DispatchData.swift
Comment (compnerd) [inline test/stdlib/DispatchData.swift]: Why the whitespace changes?

File: validation-test/Sanitizers/tsan-inout.swift
Comment (compnerd) [inline validation-test/Sanitizers/tsan-inout.swift]: This doesn't account for Windows

File: validation-test/Sanitizers/tsan-inout.swift
Comment (compnerd) [inline validation-test/Sanitizers/tsan-inout.swift]: Windows doesn't have pthreads

File: validation-test/Sanitizers/tsan-inout.swift
Comment (compnerd) [inline validation-test/Sanitizers/tsan-inout.swift]: Same

**Review by compnerd — DISMISSED** [review]
This is most certainly going to break windows.  There is a community CI up now, you should be able to do a one-off run on there.  Can you add notes to the tests that you marked as unsupported on Linux as to why they are unsupported?

File: validation-test/Sanitizers/tsan-inout.swift
Comment (yln) [reply validation-test/Sanitizers/tsan-inout.swift]: What is the general state of TSan for Swift on Windows? I was unaware that TSan tests ran at all under Windows.

**Comment by yln** [comment]
@swift-ci Please test

**Comment by yln** [comment]
@swift-ci Please test

**Comment by yln** [comment]
> This is most certainly going to break windows. There is a community CI up now, you should be able to do a one-off run on there. Can you add notes to the tests that you marked as unsupported on Linux as to why they are unsupported?

Ah, I missed that before. Can you give/point me to the instructions for doing this?

**Comment by yln** [comment]
@swift-ci Please test

**Comment by yln** [comment]
@swift-ci Please test

File: test/Sanitizers/tsan-norace-block-release.swift
Comment (yln) [reply test/Sanitizers/tsan-norace-block-release.swift]: This test didn't run on Windows before and will not run after my change.

File: test/Sanitizers/tsan-norace-deinit-run-time.swift
Comment (yln) [reply test/Sanitizers/tsan-norace-deinit-run-time.swift]: This test didn't run on Windows before and will not run after my change.

File: validation-test/Sanitizers/tsan-inout.swift
Comment (yln) [reply validation-test/Sanitizers/tsan-inout.swift]: This test didn't run on Windows before and will not run after my change.

File: test/lit.cfg
Comment (yln) [reply test/lit.cfg]: Good point. I will try to make this more resilient.

File: test/lit.cfg
Comment (yln) [reply test/lit.cfg]: I am not sure which parts exactly I have to touch for this. Can you point me to an existing variable/configuration option where this is done?

**Comment by yln** [comment]
@swift-ci Please test

**Comment by yln** [comment]
@swift-ci Please test

**Comment by yln** [comment]
@gottesmm @jrose-apple
The previous CI run succeeded. Assuming the current one does too, is this good to go?

**Comment by davezarzycki** [comment]
Since @compnerd mentioned me in this pull request, I just did a clean build of this change using a unified (not `build-script`) setup (Red Hat Fedora 30), and `ninja check-swift-validation` passes. That being said, I'm not trying to test TSan.

**Comment by yln** [comment]
@swift-ci Please test

**Comment by yln** [comment]
@swift-ci Please test

File: test/IRGen/tsan-attributes.swift
Comment (compnerd) [reply test/IRGen/tsan-attributes.swift]: Sure, I can try do a run.  BTW, @shahmishal should be able to help you get a run; there is the community CI doing windows builds, so it should be possible to do a one off.

File: validation-test/Sanitizers/tsan-inout.swift
Comment (compnerd) [reply validation-test/Sanitizers/tsan-inout.swift]: Well, all the tests run on Windows, and this is changing the fact that the test should fail to should succeed, but it wont actually do so on Windows.  Marking the TSAN tests as XFAIL on `OS=windows-msvc` is reasonable.

File: test/IRGen/tsan-attributes.swift
Comment (yln) [reply test/IRGen/tsan-attributes.swift]: Okay, I contacted Mishal. Waiting for instructions from him.

File: validation-test/Sanitizers/tsan-inout.swift
Comment (yln) [reply validation-test/Sanitizers/tsan-inout.swift]: I *think* that `// REQUIRES: tsan_runtime` makes this test not run on Windows (since we don't produce a TSan runtime there). If we start to support TSan on Windows, we have to adapt this test, but I consider this out of scope for this PR.

**Comment by compnerd** [comment]
I just did a run of this on Windows and only the following test fails:

Swift(windows-x86_64) :: IRGen/tsan_coroutines.swift

File: test/IRGen/tsan_coroutines.swift
Comment (compnerd) [inline test/IRGen/tsan_coroutines.swift]: This needs a check for `dllexport` on Windows.

File: test/lit.cfg
Comment (compnerd) [reply test/lit.cfg]: I recently wired libxml2 down.  Look at the CMakeList.txt in the test subdirectory.

Comment (yln) [reply test/lit.cfg]: I am looking at [1] but I am still unsure how to tackle this. Note that the libdispatch that we are referencing here is not built as part of the Swift cmake build, but separately through build-script.
[1] https://github.com/apple/swift/commit/a37a18f16f7211238a50f8606e76dbaee9b60712

File: test/IRGen/tsan_coroutines.swift
Comment (yln) [reply test/IRGen/tsan_coroutines.swift]: Fixed.

**Comment by yln** [comment]
@swift-ci Please test

**Comment by yln** [comment]
@swift-ci Please test

**Comment by yln** [comment]
@swift-ci Please test

File: test/Sanitizers/tsan-libdispatch.swift
Comment (jrose-apple) [inline test/Sanitizers/tsan-libdispatch.swift]: What's wrong with tvOS but not watchOS or iOS?

File: test/lit.cfg
Comment (jrose-apple) [inline test/lit.cfg]: Nitpick: since Python local variables aren't scoped, can you stick `libdispatch_` on the front of `source_dir` and `swift_module_dir` too?

File: test/lit.cfg
Comment (jrose-apple) [inline test/lit.cfg]: This still seems wacky. Why rule out TSan-without-libdispatch?

File: test/lit.cfg
Comment (jrose-apple) [inline test/lit.cfg]: Inconsistent indentation here (although really it's the other way around; @compnerd accidentally snuck some tabs into this file).

**Review by jrose-apple — COMMENTED** [review]
Sorry to delay on this. This doesn't have to block merging, but I don't love `%import-libdispatch`. What do you all think about adding these search paths unconditionally on Linux?

File: test/lit.cfg
Comment (yln) [reply test/lit.cfg]: TSan now requires libdispatch, i.e., the TSan runtime has a dependency on it. On MacOS we always have at least the system-provided libdispatch. On Linux, we need to make sure it's available.

In addition, we need to make sure that libdispatch header/swiftmodule/library for linking is available when the test itself uses it. This is what `%import-libdispatch` attempts to do.

File: test/lit.cfg
Comment (yln) [reply test/lit.cfg]: I will do that. My logic was: it's inside the conditional (which is about libdispatch) so I thought it's redundant.

File: test/lit.cfg
Comment (yln) [reply test/lit.cfg]: Fixed.

File: test/lit.cfg
Comment (jrose-apple) [reply test/lit.cfg]: Not in Python. :-(

File: test/Sanitizers/tsan-libdispatch.swift
Comment (yln) [reply test/Sanitizers/tsan-libdispatch.swift]: I included it because the other TSan tests, e.g., tsan.swift, tsan-emptyarraystorage.swift, ..., had it.

Comment (jrose-apple) [reply test/Sanitizers/tsan-libdispatch.swift]: Hm, okay.

**Comment by yln** [comment]
> ... I don't love `%import-libdispatch`.

I feel the same. The reason for it being like this is that it affects a number of substitutions, e.g., `%target-swift-emit-*`, `%target-swift-frontend`, `%target-run`, `%target-swiftc_driver` and the current approach meant a smaller, more manageable change.

File: test/lit.cfg
Comment (yln) [reply test/lit.cfg]: Actually, I can't just fix the added line:
```
  File "/home/yln/work/swift-source/swift/test/lit.cfg", line 1335
    if 'use_os_stdlib' not in lit_config.params:
    ^
IndentationError: unexpected indent
```
We can reformat the whole block/section, but I would like to do this in a separate commit.

**Comment by yln** [comment]
@swift-ci Please test

**Comment by yln** [comment]
@swift-ci Please test

File: test/lit.cfg
Comment (jrose-apple) [reply test/lit.cfg]: My point is that you should use tabs for now, sincee Saleem's code uses tabs.

**Comment by yln** [comment]
Force push to squash commit with uninteresting changes:
1) Align indentation with surrounding code/use tabs.
2) Rename variables.

@swift-ci Please test

**Comment by yln** [comment]
@swift-ci Please test

File: test/lit.cfg
Comment (yln) [reply test/lit.cfg]: We are now wiring the path through CMake!

**Comment by yln** [comment]
@swift-ci Please test

**Comment by yln** [comment]
@swift-ci Please test

**Comment by yln** [comment]
@compnerd @gottesmm @jrose-apple 
I hope I have sufficiently addressed your feedback.

The Swift test suite passes. One lldb test timed out. Can we land this?
```
Timed Out Tests (1):
    lldb-Suite :: lang/swift/stepping/TestSwiftStepping.py
```

**Comment by yln** [comment]
@swift-ci Please test OS X platform

**Comment by yln** [comment]
@swift-ci Please test

**Comment by yln** [comment]
Rebased on top of master.
@swift-ci Please test Linux platform

**Comment by yln** [comment]
@compnerd Can you take a final quick look at this?

**Comment by gottesmm** [comment]
LGTM if @compnerd is cool. I just pinged him.

**Review by compnerd — APPROVED** [review]

**Pull request merged**
